### PR TITLE
Add SyncDataPackContents and TagsLoaded lifecycle events

### PR DESCRIPTION
**Pull request by Technici4n**
This PR adds two useful lifecycle events.

1. `SyncDataPackContents`. Quite straightforward. I need this event for Modern Dynamics where I have a custom resource reload listener whose contents I want to sync with the client, to display in REI. A similar effect could be achieved using a connection event, and the EndDataPackReload event, however the latter runs too late if REI expects all data to have arrived before the tag and the recipe sync packets.
2. `TagsLoaded`. This event is useful when a tag-dependent cache needs to be invalidated because tags have changed. Not sure if the `DynamicRegistryManager` and the `isClient` parameters are useful, but I thought I might as well provide them. There were two places in Fabric API that I converted to use this event (`FuelRegistry` cache, and the `MiningLevelManagerImpl`). It also turns out that `FuelRegistry` didn't invalidate its cache when a server is started, which could theoretically be an issue when logging out of a singleplayer and then logging back in with a different datapack.

### CODE REVIEW

**Comment by Technici4n** [comment]
Interesting prior art for `TagsLoaded`: (mojmap ofc) https://github.com/MinecraftForge/MinecraftForge/pull/8636, I just realized.

File: fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/CommonLifecycleEvents.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			callback.tagsLoaded(registries, isClient);
+		}
+	});
+
+	public interface TagsLoaded {
+		/**
+		 * @param registries Up-to-date registries from which the tags can be retrieved.
+		 * @param isClient True if the client just received a sync packet, false if the server just (re)loaded the tags.
+		 */
+		void tagsLoaded(DynamicRegistryManager registries, boolean isClient);
```
Comment (Juuxel) [inline fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/CommonLifecycleEvents.java]: This should have a more method-y name. Also, `is` prefix for params goes against yarn a bit

File: fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java [hunk 2]
```diff
@@ -116,6 +129,41 @@ public interface ServerStopped {
...
+		 *     Packet<?> packet = ServerPlayNetworking.createS2CPacket(CHANNEL_NAME, buf);
+		 *
+		 *     // 2) Send to relevant player(s)
+		 *     if (player == null) {
+		 *         // Reload: send to all
+		 *         playerManager.sendToAll(packet);
+		 *     } else {
+		 *         // Player joined: send only to that player
+		 *         player.networkHandler.sendPacket(packet);
+		 *     }
```
Comment (Juuxel) [inline fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java]: This feels weird, like there's two events in one.

Comment (Technici4n) [reply fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java]: Alternatives:
- Having two events.
- Sending the event for each player, and needing to create a new packet every time.

I don't think the alternatives are particularly nicer than this, especially with the extensive javadoc explaining how to use the event.

**Comment by Technici4n** [comment]
Since this is a prerequisite for a clean fix of #2108, I am marking this as medium priority.

File: fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java (hunk 2 above)
Comment (Technici4n) [reply fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerLifecycleEvents.java]: Changed to sending the event for each player, the API shape is a lot better, and I think trying to not serialize the packet multiple times is premature optimization. If this proves to be a bottleneck, we can always introduce a specialized event later on.

**Review by Player3324 — APPROVED** [review]

**Review by modmuss50 — APPROVED** [review]

**Review by Juuxel — APPROVED** [review]

**Pull request merged**